cpu/rpx0xx/periph/spi.c

```diff
@@ -0,0 +1,413 @@
+        gpio_set_io_config(spi_config[spi].mosi_pin, mosi_io_config);
+    }
+    if (gpio_is_valid(spi_config[spi].clk_pin)) {
+        gpio_set_pad_config(spi_config[spi].clk_pin, clk_pad_config);
+        gpio_set_io_config(spi_config[spi].clk_pin, clk_io_config);
+    }
+    if (gpio_is_valid(spi_config[spi].miso_pin)) {
+        gpio_set_pad_config(spi_config[spi].miso_pin, miso_pad_config);
+        gpio_set_io_config(spi_config[spi].miso_pin, miso_io_config);
+    }
+}
+
+void spi_deinit_pins(spi_t spi)
+{
+    DEBUG("[rpx0xx] Call spi_deinit_pins(spi=%" PRIdFAST8 ")\n", spi);
+
+    assert((unsigned)spi < SPI_NUMOF);
+
+    if (gpio_is_valid(spi_config[spi].mosi_pin)) {
+        gpio_reset_all_config(spi_config[spi].mosi_pin);
+    }
+
+    if (gpio_is_valid(spi_config[spi].miso_pin)) {
+        gpio_reset_all_config(spi_config[spi].miso_pin);
+    }
+
+    if (gpio_is_valid(spi_config[spi].clk_pin)) {
+        gpio_reset_all_config(spi_config[spi].clk_pin);
+    }
+}
+
+int spi_init_cs(spi_t spi, spi_cs_t cs)
+{
+    DEBUG("[rpx0xx] Call spi_init_cs(spi=%" PRIdFAST8 ", cs=%" PRIu32 ")\n",
+          spi, cs);
+
+    if (spi >= SPI_NUMOF) {
+        return SPI_NODEV;
+    }
+    if (!gpio_is_valid(cs)) {
+        return SPI_NOCS;
+    }
+
+    gpio_init((gpio_t)cs, GPIO_OUT);
+    gpio_set((gpio_t)cs);
+
+    return SPI_OK;
+}
+
+int spi_deinit_cs(spi_t spi, spi_cs_t cs)
+{
+    DEBUG("[rpx0xx] Call spi_deinit_cs(spi=%" PRIdFAST8 ", cs=%" PRIu32 ")\n",
+          spi, cs);
+
+    if (spi >= SPI_NUMOF) {
+        return SPI_NODEV;
+    }
+    if (!gpio_is_valid(cs)) {
+        return SPI_NOCS;
+    }
+
+    gpio_reset_all_config((gpio_t)cs);
+
+    return SPI_OK;
+}
+
+static inline void _check_best_clk_result(uint16_t target_dvsr,
+                                          uint16_t cur_dvsr,
+                                          uint16_t scr,
+                                          uint8_t cpsdvsr,
+                                          uint16_t *best_diff,
+                                          uint16_t *best_scr,
+                                          uint8_t *best_cpsdvsr)
+{
+    uint16_t diff = (target_dvsr > cur_dvsr)
+                    ? target_dvsr - cur_dvsr
+                    : cur_dvsr - target_dvsr;
+
+    if (diff < *best_diff) {
+        *best_diff = diff;
+        *best_cpsdvsr = cpsdvsr;
+        *best_scr = scr;
+    }
+}
+
+static void _calc_pl022_clk(_pl022_clk_t *pl022_clk, spi_clk_t clk)
+{
+    uint32_t clk_hz;
+
+    switch (clk) {
+    case SPI_CLK_100KHZ:
+        clk_hz = KHZ(100);
+        break;
+    case SPI_CLK_400KHZ:
+        clk_hz = KHZ(400);
+        break;
+    case SPI_CLK_1MHZ:
+        clk_hz = MHZ(1);
+        break;
+    case SPI_CLK_5MHZ:
+        clk_hz = MHZ(5);
+        break;
+    case SPI_CLK_10MHZ:
+        clk_hz = MHZ(10);
+        break;
+    default:
+        clk_hz = KHZ(100);
+        break;
+    }
```

Comment:
Instead of this, just do

```C
#define HAVE_SPI_CLK_T
enum {
    SPI_CLK_100KHZ = KHZ(100), /**< drive the SPI bus with 100KHz */
    SPI_CLK_400KHZ = KHZ(400), /**< drive the SPI bus with 400KHz */
    SPI_CLK_1MHZ   = MHZ(1),   /**< drive the SPI bus with 1MHz */
    SPI_CLK_5MHZ   = MHZ(5),   /**< drive the SPI bus with 5MHz */
    SPI_CLK_10MHZ  = MHZ(10),  /**< drive the SPI bus with 10MHz */
};
```

in `periph_cpu.h`

Reply:
I am not familiar with  this kind of statement. Is `clk` guaranteed to be 32 bits wide with this statement?

How about this instead?
```
const uint32_t CLK_MAP[] = {
    [SPI_CLK_100KHZ] KHZ(100),
    [SPI_CLK_400KHZ] KHZ(400),
    [SPI_CLK_1MHZ]   MHZ(1),
    [SPI_CLK_5MHZ]   MHZ(5),
    [SPI_CLK_10MHZ]  MHZ(10),
};
uint32_t clk_hz = CLK_MAP[clk];
```

Some assertion is needed anyway to ensure `clk`has a valid value.

Reply:
The idea is to allow arbitrary clock frequencies, not just the pre-defined values.
The enum will automatically assume the needed width.

Reply:
> See e.g. [`cpu/gd32v`](https://github.com/RIOT-OS/RIOT/blob/master/cpu/gd32v/include/periph_cpu.h#L324)

Ah ok, now I got it! Sure, that makes sense.